internal/config/node/cgroups.go

```diff
@@ -0,0 +1,128 @@
+// +build linux
+
+package node
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/containers/libpod/pkg/cgroups"
+	"github.com/pkg/errors"
+)
+
+var (
+	cgroupIsV2Once sync.Once
+	cgroupIsV2     bool
+	cgroupIsV2Err  error
+
+	cgroupHasMemorySwapOnce sync.Once
+	cgroupHasMemorySwap     bool
+	cgroupHasMemorySwapErr  error
+
+	cgroupControllerOnce sync.Once
+	cgroupControllerErr  error
+	cgroupHasHugetlb     bool
+	cgroupHasPid         bool
+)
+
+// CgroupIsV2 returns whether we are using cgroup v2 or v1
+func CgroupIsV2() bool {
+	cgroupIsV2Once.Do(func() {
+		unified, err := cgroups.IsCgroup2UnifiedMode()
+		cgroupIsV2 = unified
+		cgroupIsV2Err = err
+	})
+	return cgroupIsV2
+}
+
+// CgroupHasMemorySwap returns whether the memory swap controller is present
+func CgroupHasMemorySwap() bool {
+	cgroupHasMemorySwapOnce.Do(func() {
+		if CgroupIsV2() {
+			cgroupHasMemorySwap = true
+			return
+		}
+
+		_, err := os.Stat("/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes")
+		if err != nil {
+			cgroupHasMemorySwapErr = errors.New("node not configured with memory swap")
+			cgroupHasMemorySwap = false
+			return
+		}
+
+		cgroupHasMemorySwap = true
+	})
+	return cgroupHasMemorySwap
+}
+
+// CgroupHasHugetlb returns whether the hugetlb controller is present
+func CgroupHasHugetlb() bool {
+	cgroupControllerOnce.Do(func() {
+		checkRelevantControllers()
+	})
+	return cgroupHasHugetlb
+}
+
+// CgroupHasPid returns whether the pid controller is present
+func CgroupHasPid() bool {
+	cgroupControllerOnce.Do(func() {
+		checkRelevantControllers()
+	})
+	return cgroupHasPid
+}
+
+type controllerInfo struct {
+	name    string
+	enabled *bool
+}
+
+func checkRelevantControllers() {
+	relevantControllers := []controllerInfo{
+		{
+			name:    "pids",
+			enabled: &cgroupHasPid,
+		},
+		{
+			name:    "hugetlb",
+			enabled: &cgroupHasHugetlb,
+		},
+	}
+
+	if CgroupIsV2() {
+		checkRelevantControllersV2(relevantControllers)
+		return
+	}
+	checkRelevantControllersV1(relevantControllers)
+}
+
+func checkRelevantControllersV1(relevantControllers []controllerInfo) {
+	allControllers, cgroupControllerErr := ioutil.ReadDir("/sys/fs/cgroup")
+	if cgroupControllerErr != nil {
+		return
+	}
+	for _, toCheck := range relevantControllers {
+		for _, controllerStat := range allControllers {
+			if toCheck.name == controllerStat.Name() && controllerStat.IsDir() {
```

Comment:
nit: in general, the shorter the variable scope, the shorter its name should be. The scope of `controllerStat` is just two lines, the code will be more readable if you call it `c` or `st`.

nit2: `IsDir()` check is cheaper, thus it should go first.

Reply:
I'll just counter the argument.  I know shorter is in many style guides, but I find it much clearer to read a variable name than just a char or two, even if it only exists across a few lines.  My $.02 and YMMV.

Reply:
I fixed the allControllers variable to be more informative, and then made the iterative have a shorter name: a compromise :)